### PR TITLE
Fixed #34362 -- Fixed FilteredRelation() crash on conditional expressions.

### PR DESCRIPTION
**Pull request by panicofr**
Thanks zhu for the report

### CODE REVIEW

File: django/db/models/sql/query.py [hunk 1]
```diff
@@ -1573,7 +1573,11 @@ def build_filtered_relation_q(
 
     def add_filtered_relation(self, filtered_relation, alias):
         filtered_relation.alias = alias
-        lookups = dict(get_children_from_q(filtered_relation.condition))
+        lookups = dict(
+            l
+            for l in get_children_from_q(filtered_relation.condition)
+            if isinstance(l, tuple)
+        )
```
Comment (charettes) [inline django/db/models/sql/query.py:1580]: I think this is only _masking_ the crash by dropping the false criteria

If it was generating a join along the lines of `LEFT JOIN book AS the_book ON (book.author_id = author.id AND false)`, as should be expected in this case, then `the_book.id` should always be `NULL` as `(book.author_id = author.id AND false)` will always be false.

Your test demonstrates that the criteria is dropped and produces wrong results.

File: django/db/models/sql/query.py [hunk 2]
```diff
@@ -1573,7 +1573,11 @@ def build_filtered_relation_q(
 
     def add_filtered_relation(self, filtered_relation, alias):
         filtered_relation.alias = alias
-        lookups = dict(get_children_from_q(filtered_relation.condition))
+        lookups = dict(
```
Comment (charettes) [inline django/db/models/sql/query.py:1576]: I think this is only _masking_ the crash by dropping the false criteria

If it was generating a join along the lines of `LEFT JOIN book AS the_book ON (book.author_id = author.id AND false)`, as should be expected in this case, then `the_book.id` should always be `NULL` as `(book.author_id = author.id AND false)` will always be false.

Your test demonstrates that the criteria is dropped and produces wrong results.

The code must be adapted to handle conditional expressions or raise a more meaningful error when provided to `FilteredRelation(condition)` not discard them.

Comment (panicofr) [reply django/db/models/sql/query.py:1576]: Hi @charettes, 

the method I changed drop the false criteria only during the check of lookups. The filtered lookup dict get discarded after the check.

This is the resulting SQL query, according to me it's the expected one:

```SELECT
    "filtered_relation_author"."id",
    "filtered_relation_author"."name",
    "filtered_relation_author"."content_type_id",
    "filtered_relation_author"."object_id"
FROM "filtered_relation_author" INNER JOIN "filtered_relation_book" the_book
ON ("filtered_relation_author"."id" = the_book."author_id" AND (False))
WHERE the_book."id" IS NOT NULL
```

Let me know if I'm missing something 😉

Comment (charettes) [reply django/db/models/sql/query.py:1576]: I realize I should have made my comment clearer and I misunderstood the test, sorry for that.

When `child` is an instance of `Expression` it could refer to any lookup though (e.g. `Exact('otherfield__outside', 3)`) so the question that remains with this ticket is whether we want to validate that as well.

Comment (panicofr) [reply django/db/models/sql/query.py:1576]: @charettes I got the point.. let me work on it 👍

Comment (panicofr) [reply django/db/models/sql/query.py]: @charettes I have implemented a fix for this case and I have added some tests to increase coverage. By the way, I found a bug when the condition didn't specify an explicit lookup.

Let me know what do you think... thanks!

**Comment by zhu** [comment]
The `rhs` in tuple did not check. I think it also need checking.
And also some expressions like: `Func`, `When`, `Subquery`/`Exists` with `OuterRef` annotated, `ExpressionWrapper`.


    def test_conditional_expression_with_case(self):
        msg = (
            "FilteredRelation's condition doesn't support relations outside "
            "the 'book__editor' (got 'book__author__name')."
        )
        with self.assertRaisesMessage(ValueError, msg):
            Author.objects.annotate(
                book_editor=FilteredRelation(
                    "book__editor",
                    condition=Q(Case(
                        When(
                            # the_book__title__startswith='P',
                            book__author__name='Alice',
                            then=True,
                        ),
                        default=False
                    )),
                ),
            )

    def test_conditional_expression_with_subquery(self):
        self.assertSequenceEqual(
            Author.objects.annotate(
                book_editor=FilteredRelation(
                    "book__editor",
                    condition=Q(Exists(
                        Editor.objects.annotate(
                            author_name=OuterRef('name'),
                        ).filter(
                            author_name__istartswith=F('name'),
                        )
                    )),
                ),
            ),
            [self.author1],
        )

    def test_conditional_expression_with_expressionwrapper(self):
        msg = (
            "FilteredRelation's condition doesn't support relations outside "
            "the 'book__editor' (got 'book__author__name')."
        )
        with self.assertRaisesMessage(ValueError, msg):
            Author.objects.annotate(
                book_editor=FilteredRelation(
                    "book__editor",
                    condition=Q(ExpressionWrapper(
                        Q(
                            Exact(F("book__author__name"), "Alice"),
                            Value(True),
                            book__title__startswith="Poem",
                        ),
                        output_field=BooleanField()
                    )),
                ),
            )

File: django/db/models/sql/query.py (hunk 2 above)
Comment (zhu) [reply django/db/models/sql/query.py]: > 

`INNER JOIN "filtered_relation_book" the_book` implicit `the_book."id" IS NOT NULL`

File: django/db/models/sql/query.py [hunk 3]
```diff
@@ -65,12 +68,35 @@ def get_field_names_from_opts(opts):
...
+    if hasattr(expr, "rhs"):
+        yield from get_paths_from_expression(expr.rhs)
+    if isinstance(expr, F):
+        yield expr.name
+    if isinstance(expr, Func):
+        for func_expression in expr.source_expressions:
+            yield from get_paths_from_expression(func_expression)
+    if isinstance(expr, Case):
+        for case in expr.cases:
+            yield from get_children_from_q(case.condition)
```
Comment (felixxm) [inline django/db/models/sql/query.py:83]: As far as I'm aware, we don't need to handle all kind of expressions separately. We should be able to reuse `Expression.get_source_expressions()`. For example the following works for me:
```python
def get_paths_from_expression(expr):
    if isinstance(expr, F):
        yield expr.name 
    elif isinstance(expr, Q):
        yield from get_children_from_q(expr)
    elif hasattr(expr, "resolve_expression"):
        for source_expr in expr.get_source_expressions():
            yield from get_paths_from_expression(source_expr)

```

File: tests/filtered_relation/tests.py [hunk 4]
```diff
@@ -685,6 +703,93 @@ def test_eq(self):
...
+                        ExpressionWrapper(
+                            Q(
+                                Value(True),
+                                Exact(F("book__author__name"), "Alice"),
+                            ),
+                            output_field=BooleanField(),
+                        )
+                    ),
+                ),
+            )
```
Comment (felixxm) [inline tests/filtered_relation/tests.py:791]: We added new tests for `ValueError`, can we also add tests with lookups, `Case()`, `ExpressionWrapper` that don't raise `ValueError`?

File: django/db/models/sql/query.py [hunk 5]
```diff
@@ -65,12 +68,35 @@ def get_field_names_from_opts(opts):
...
             yield from get_children_from_q(child)
         else:
-            yield child
+            if isinstance(child, tuple):
+                lhs, rhs = child
+                yield lhs
+                if hasattr(rhs, "resolve_expression"):
+                    yield from get_paths_from_expression(rhs)
+            if isinstance(child, ExpressionWrapper):
+                yield from get_children_from_q(child.expression)
```
Comment (felixxm) [inline django/db/models/sql/query.py:97]: This is unnecessary with the simplified version of `get_paths_from_expression()` that handles `Q`:
```suggestion
```

**Review by felixxm — COMMENTED** [review]
@panicofr Thanks :+1: I wonder if we could use `Q.flatten()` or `Expression.flatten()` somewhere here :thinking: 

I rebased from the `main` branch to get rid of unrelated commits.

**Comment by felixxm** [comment]
> @panicofr Thanks +1 I wonder if we could use `Q.flatten()` or `Expression.flatten()` somewhere here.

Can you comment on it somehow? :point_up: Is it doable? If no, why? Thanks :+1:

File: django/db/models/sql/query.py [hunk 6]
```diff
@@ -65,12 +65,33 @@ def get_field_names_from_opts(opts):
...
+        return
+
+    children = [
+        child_expr for child_expr in expr.flatten() if isinstance(child_expr, (Q, F))
+    ]
+    for child_expr in children:
+        if isinstance(child_expr, F):
+            yield child_expr.name
+        elif isinstance(child_expr, Q):
+            yield from get_children_from_q(child_expr)
```
Comment (felixxm) [inline django/db/models/sql/query.py:80]: Do we need a temporary list? Maybe:
```suggestion
    if isinstance(expr, F):
        yield expr.name
    elif hasattr(expr, "flatten"):
        for child in expr.flatten():
            if isinstance(child, F):
                yield child_expr.name
            elif isinstance(child, Q):
                yield from get_children_from_q(child_expr)
```

or even
```python
def get_paths_from_expression(expr):
    if isinstance(expr, F):
        yield expr.name
    elif hasattr(expr, "flatten"):
        for child in expr.flatten():
            if isinstance(child, Q):
                yield from get_children_from_q(child)
            else:
                yield from get_paths_from_expression(child)
```

**Comment by felixxm** [comment]
@zhu Does it work for you?

**Comment by zhu** [comment]
> @zhu Does it work for you?

It seems fine except for `test_conditional_expression_with_subquery`.

**Comment by felixxm** [comment]
> > @zhu Does it work for you?
> 
> It seems fine except for `test_conditional_expression_with_subquery`.

Thanks for checking :+1: `test_conditional_expression_with_subquery` is a separate issue, see ticket-31714.

File: django/db/models/expressions.py [hunk 7]
```diff
@@ -1367,11 +1367,20 @@ def resolve_expression(
...
         c.is_summary = summarize
         if hasattr(c.condition, "resolve_expression"):
+            extra_args = (
+                dict(reuse_with_filtered_relation=True)
+                if isinstance(c.condition, Q)
+                else {}
+            )
             c.condition = c.condition.resolve_expression(
-                query, allow_joins, reuse, summarize, False
+                query, allow_joins, reuse, summarize, False, **extra_args
```
Comment (felixxm) [inline django/db/models/expressions.py:1376]: I wonder why we only pass `reuse_with_filtered_relation` here :thinking:. It seems that we're only fixing an issue in `When()`, but in theory I could appear in other places. This looks especially suspicious since we're not even aware that this `When()` is used with `FilteredRelation()` or not. Maybe we always should pass `reuse_with_filtered_relation=True` to `_add_q()` when resolving `Q` objects :thinking: (\cc @charettes).

Comment (panicofr) [reply django/db/models/expressions.py]: I agree, I've been mislead because `When()` is the only expression in tests having a `Q()` object as child.

I like this approach, cleaner and more generic

File: django/db/models/query_utils.py [hunk 8]
```diff
@@ -91,6 +91,7 @@ def resolve_expression(
             split_subq=False,
             check_filterable=False,
             summarize=summarize,
+            reuse_with_filtered_relation=True,
```
Comment (charettes) [inline django/db/models/query_utils.py:94]: This seems wrong as it will now make any `Q` resolving take the `SQLQuery.join` path dedicated to dealing with filtered relations.

https://github.com/django/django/blob/01ae9d4ca9afdaf30a247e10e8333261a7d8224c/django/db/models/sql/query.py#L997-L1016

File: django/db/models/sql/query.py [hunk 9]
```diff
@@ -1936,7 +1959,11 @@ def resolve_ref(self, name, allow_joins=True, reuse=None, summarize=False):
                     annotation = self.try_transform(annotation, transform)
                 return annotation
             join_info = self.setup_joins(
-                field_list, self.get_meta(), self.get_initial_alias(), can_reuse=reuse
+                field_list,
+                self.get_meta(),
+                self.get_initial_alias(),
+                can_reuse=reuse,
+                reuse_with_filtered_relation=True,
```
Comment (charettes) [inline django/db/models/sql/query.py:1966]: Ditto about systematically taking the logic meant to only run when resolving filtered relations.

**Review by charettes — COMMENTED** [review]
I think that systematically passing `reuse_with_filtered_relation=True` could have unintended side effects as it will make `join` and `setup_join` always operate in a mode that was designed to only run when doing the unfortunate deferred resolving that happens in `FilteredRelation.as_sql`.

I'm afraid we won't be able to address this issue without tackling ticket-33766 to address this unfortunate sequence of calls.

What we need here to address both ticket is to have the resolving of `FilteredRelation.condition` happen immediately when its referenced the first time in `sql.Query.add_filtered_relation`.

**Comment by charettes** [comment]
@felixxm, @panicofr here's an example of what I had in mind by that https://github.com/django/django/pull/16786

**Comment by felixxm** [comment]
@panicofr Please rebase after d660cee5bc68b597503c2a16f3d9928d52f93fb4.

**Comment by panicofr** [comment]
@felixxm done 👍 

@charettes Thanks for the your work!

I think that `filtered_relation.tests.FilteredRelationTests.test_conditional_expression_with_func` expected behavior (failing at the moment) conflicts with other tests like `filtered_relation.tests.FilteredRelationAggregationTests.test_condition_spans_join_chained`. 

Probably this test should not raise a `ValueError`, the output SQL seems acceptable 🤔

**Comment by charettes** [comment]
@panicofr I see what you mean.

The difference between the two tests is that `test_condition_spans_join_chained` (and `test_condition_spans_join` for that matter) is that while their `condition` do spawn joins outside of the `relation` prefix (`editor`) none of their parts intersect with the joins required to materialize the `relation`. This distinction is what avoids the ambiguity the towards joins reuse.

In reality it's multi-valued relationships that usually cause problems from a reusability perspective and `test_conditional_expression_with_func` clearly demonstrates that proper SQL is generated when they are not involved (a book has a single editor and author) but there are other cases that I suspect are still breaking (e.g. see ticket-34229)

I'm not opposed to relaxing this check in a follow up MR if we can demonstrate it does the right thing with the recent changes in this area but since this ticket is focused on conditional expression support I would instead adapt `test_conditional_expression_with_func` not to use this pattern.

**Comment by panicofr** [comment]
@charettes yes, it makes sense. I added a check to test if condition path intersects relation joins. If not, no `ValueError` is raised in case of diverging paths. Let me know if it could work for you.

Thanks!

File: django/db/models/sql/query.py [hunk 10]
```diff
@@ -65,22 +65,46 @@ def get_field_names_from_opts(opts):
...
+def get_child_with_renamed_prefix(prefix, replacement, child):
+    if isinstance(child, Node):
+        return rename_prefix_from_q(prefix, replacement, child)
+    if isinstance(child, tuple):
+        lhs, rhs = child
+        lhs = lhs.replace(prefix, replacement, 1)
+        if hasattr(rhs, "resolve_expression"):
+            rhs = get_child_with_renamed_prefix(prefix, replacement, rhs)
+        return (lhs, rhs)
+    return child.copy()
```
Comment (charettes) [inline django/db/models/sql/query.py:102]: Pretty sure you need to handle the case where a child has a `resolve_expression` as well. You'll likely need to rely on `get_source_expressions` and `set_source_expressions`

For example, `Q(GreaterThan(F("books__author__age"), F("books__author__num_books")))` with a prefix change of `books__author -> filtered_authors` should be turned into `Q(GreaterThan(F("filtered_authors__age"), F("filtered_authors__num_books")))`.

Currently only a copy is returned.

File: django/db/models/sql/query.py [hunk 11]
```diff
@@ -1581,13 +1604,17 @@ def add_filtered_relation(self, filtered_relation, alias):
...
-            shift = 2 if not lookup_parts else 1
-            lookup_field_path = lookup_field_parts[:-shift]
+        for lookup in get_children_from_q(filtered_relation.condition):
+            _, lookup_field_parts, _ = self.solve_lookup_type(lookup)
+            lookup_field_path = lookup_field_parts[:-1]
+            intersects = (
+                lookup_field_path
+                and relation_field_parts
+                and lookup_field_path[0] == relation_field_parts[0]
+            )
```
Comment (charettes) [inline django/db/models/sql/query.py:1614]: Not entirely sold on the necessity of this change in this PR.

What I meant by https://github.com/django/django/pull/16622#issuecomment-1529061058 is that `test_conditional_expression_with_func` should be adapted not to require adjusting this logic.

In other words, I believe there are ways to test that `FilteredRelation` doesn't _crash_ on conditional expressions (which is the purpose of this PR) without getting into adjusting what constitutes a violation? Maybe I'm missing something though.

Comment (panicofr) [reply django/db/models/sql/query.py:1614]: @charettes Thanks for the explanation. We could solve this is avoiding to check paths in the rhs of expressions, and changing `test_conditional_expression_with_func` to not assumine that queryset evaluation raise a `ValueError`.

Checking rhs  was the matter of [this comment](https://github.com/django/django/pull/16622#issuecomment-1455677532) by @zhu , but probably we could postpone this in a future PR, as you suggested.

File: tests/filtered_relation/tests.py [hunk 12]
```diff
@@ -629,6 +634,15 @@ def test_condition_outside_relation_name(self):
...
             )
 
+    def test_condition_with_exact_lookup_outside_relation_name(self):
+        qs = Author.objects.annotate(
+            book_editor=FilteredRelation(
+                "book__editor",
+                condition=Q(book__author__name="book"),
+            ),
+        ).filter(book_editor__isnull=True)
+        self.assertEqual(qs.count(), 4)
```
Comment (felixxm) [inline tests/filtered_relation/tests.py:644]: This test and `test_conditional_expression_with_func` work without this patch so I moved them to a separate PR. Both are also quite surprising especially since `test_condition_outside_relation_name` raises an exception, but this is a discussion for a separate ticket.

Comment (felixxm) [reply tests/filtered_relation/tests.py:644]: Check out #17100.

**Comment by felixxm** [comment]
Rebased.

File: tests/filtered_relation/tests.py [hunk 13]
```diff
@@ -707,6 +710,139 @@ def test_eq(self):
...
+    def test_conditional_expression_with_lookup(self):
+        qs = (
+            Author.objects.annotate(
+                poem_book=FilteredRelation(
+                    "book",
+                    condition=Q(IStartsWith(F("book__title"), "poem")),
+                ),
+            )
+            .filter(poem_book__isnull=False)
+            .distinct()
```
Comment (felixxm) [inline tests/filtered_relation/tests.py:786]: `distinct()` should not be needed :thinking: As far as I'm aware,
```
condition=Q(IStartsWith(F("book__title"), "poem"))
``` 
should work exactly the same as
```
condition=Q(book__title__istartswith="poem")
```
and returns a singe author. It is also suspicious to use `distinct()` in other tests.

File: django/db/models/sql/query.py [hunk 14]
```diff
@@ -65,20 +65,58 @@ def get_field_names_from_opts(opts):
...
+            yield lhs
+            if hasattr(rhs, "resolve_expression"):
+                yield from get_paths_from_expression(rhs)
+        elif hasattr(child, "resolve_expression"):
+            yield from get_paths_from_expression(child)
 
 
-def get_child_with_renamed_prefix(prefix, replacement, child):
+def replace_child_name(child_name, prefix, replacement):
+    return child_name.replace(prefix, replacement, 1)
```
Comment (felixxm) [inline django/db/models/sql/query.py:93]: It's one-liner, IMO, it's not worth extra function calls. I'd move this logic directly to the `get_child_with_renamed_prefix()`.

File: tests/filtered_relation/tests.py [hunk 15]
```diff
@@ -707,6 +710,129 @@ def test_eq(self):
...
+                    ExpressionWrapper(
+                        Q(
+                            Exact(F("book__title"), "Poem by Alice"),
+                        ),
+                        output_field=BooleanField(),
+                    )
+                ),
+            ),
+        ).filter(poem_book__isnull=False)
+        self.assertEqual(qs.count(), 1)
```
Comment (felixxm) [inline tests/filtered_relation/tests.py:804]: I would change all `count()` assertions to checking model instances.

File: django/db/models/sql/query.py [hunk 16]
```diff
@@ -65,20 +65,58 @@ def get_field_names_from_opts(opts):
...
     if isinstance(child, Node):
         return rename_prefix_from_q(prefix, replacement, child)
-    lhs, rhs = child
-    lhs = lhs.replace(prefix, replacement, 1)
-    return lhs, rhs
+    if isinstance(child, tuple):
+        lhs, rhs = child
+        lhs = replace_child_name(lhs, prefix, replacement)
+        if hasattr(rhs, "resolve_expression"):
+            rhs = get_child_with_renamed_prefix(prefix, replacement, rhs, is_lhs=False)
```
Comment (felixxm) [inline django/db/models/sql/query.py:103]: What do you think about skipping `F` instances instead of adding an extra kwarg?, e.g.
```diff
diff --git a/django/db/models/sql/query.py b/django/db/models/sql/query.py
index 2634a33d98..1b476a6dcf 100644
--- a/django/db/models/sql/query.py
+++ b/django/db/models/sql/query.py
@@ -93,19 +93,18 @@ def replace_child_name(child_name, prefix, replacement):
     return child_name.replace(prefix, replacement, 1)
 
 
-def get_child_with_renamed_prefix(prefix, replacement, child, is_lhs=True):
+def get_child_with_renamed_prefix(prefix, replacement, child):
     if isinstance(child, Node):
         return rename_prefix_from_q(prefix, replacement, child)
     if isinstance(child, tuple):
         lhs, rhs = child
         lhs = replace_child_name(lhs, prefix, replacement)
-        if hasattr(rhs, "resolve_expression"):
-            rhs = get_child_with_renamed_prefix(prefix, replacement, rhs, is_lhs=False)
+        if not isinstance(rhs, F) and hasattr(rhs, "resolve_expression"):
+            rhs = get_child_with_renamed_prefix(prefix, replacement, rhs)
         return lhs, rhs
     if isinstance(child, F):
         child = child.copy()
-        if is_lhs:
-            child.name = replace_child_name(child.name, prefix, replacement)
+        child.name = replace_child_name(child.name, prefix, replacement)
         return child
     if hasattr(child, "resolve_expression"):
         child = child.copy()

```

**Review by felixxm — COMMENTED** [review]
@panicofr Thanks for updates :+1:

File: tests/filtered_relation/tests.py [hunk 17]
```diff
@@ -707,6 +710,139 @@ def test_eq(self):
...
+                        Value(True),
+                        book__title__startswith="Poem",
+                    ),
+                ),
+            ).filter(book_editor__isnull=False),
+            [self.author1],
+        )
+
+    def test_conditional_expression_with_case(self):
+        self.assertSequenceEqual(
```
Comment (felixxm) [inline tests/filtered_relation/tests.py:737]:
```suggestion
        self.assertCountEqual(
```

**Comment by felixxm** [comment]
@panicofr Thanks :+1: I removed unused `is_lhs` kwarg, and simplified tests a bit.
